[folly related] update to 04.21

c8ef opened the pull request:
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md).
- [x] SHA512s are updated for each updated download.
- [x] The "supports" clause reflects platforms that may be fixed by this new version.
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

c8ef committed:
[folly related] update to 04.21

c8ef committed:
[folly related] update to 04.21

c8ef committed:
[folly related] update to 04.21

c8ef committed:
[folly related] update to 04.21

c8ef commented:
I'm wondering if the CI rule has changed in any way? For example, in `arm_neon_android`, both this week's and last week's CI had the same message, but this week's CI failed.🧐

```
folly[core]:arm-neon-android only supports (windows & x64 & !uwp & !mingw) | (!windows & (x64 | arm64))
fizz[core]:arm-neon-android resulted in the unexpected state cascade because the following dependencies did not build: folly[core]:arm-neon-android only supports (windows & x64 & !uwp & !mingw) | (!windows & (x64 | arm64))
wangle[core]:arm-neon-android resulted in the unexpected state cascade because the following dependencies did not build: folly[core]:arm-neon-android only supports (windows & x64 & !uwp & !mingw) | (!windows & (x64 | arm64))
mvfst[core]:arm-neon-android resulted in the unexpected state cascade because the following dependencies did not build: folly[core]:arm-neon-android only supports (windows & x64 & !uwp & !mingw) | (!windows & (x64 | arm64))
proxygen[core]:arm-neon-android resulted in the unexpected state cascade because the following dependencies did not build: folly[core]:arm-neon-android only supports (windows & x64 & !uwp & !mingw) | (!windows & (x64 | arm64))
```

dg0yt commented:
CI now test each individual feature. Add expected cascades to `scripts/ci.features.baseline.txt`.

BillyONeal commented:
> I'm wondering if the CI rule has changed in any way? For example, in `arm_neon_android`, both this week's and last week's CI had the same message, but this week's CI failed.🧐

This is because we accidentally made the exit code 0 even when there were failures 😳 https://github.com/microsoft/vcpkg-tool/pull/1649

c8ef commented:
> > I'm wondering if the CI rule has changed in any way? For example, in `arm_neon_android`, both this week's and last week's CI had the same message, but this week's CI failed.🧐
> 
> This is because we accidentally made the exit code 0 even when there were failures 😳 [microsoft/vcpkg-tool#1649](https://github.com/microsoft/vcpkg-tool/pull/1649)

Okay, but since I pulled the latest commit, shouldn't the vcpkg tool on this branch be using the latest version?

BillyONeal commented:
> Okay, but since I pulled the latest commit, shouldn't the vcpkg tool on this branch be using the latest version?

Even if it wasn't in your branch, GitHub tests a merge of your branch with the target branch. It wasn't red *last time* because at the time the exit code bug was still there.

c8ef commented:
Oh, I got it. The newly updated tool unveils a hidden bug. I will try to resolve these failures tonight.

c8ef committed:
[folly related] update to 04.21

c8ef committed:
[folly related] update to 04.21

c8ef committed:
[folly related] update to 04.21

c8ef committed:
[folly related] update to 04.21

c8ef committed:
[folly related] update to 04.21

c8ef committed:
[folly related] update to 04.21

c8ef commented:
The CI looks good now. Note that all these packages depend on folly, so the platform they support should not exceed what folly supports. I have adjusted the support platform for these packages.

Please take another look~ @LilyWangLL

dg0yt commented:
> CI now test each individual feature. Add expected cascades to `scripts/ci.features.baseline.txt`.

I really meant updating only this file, not adding "supports" everywhere...
It is a pure CI issue ATM.

c8ef commented:
> > CI now test each individual feature. Add expected cascades to `scripts/ci.features.baseline.txt`.
> 
> I really meant updating only this file, not adding "supports" everywhere... It is a pure CI issue ATM.

But I don't think it makes sense for A to rely on B when A's build scope is bigger than B. We can gradually expand the supported platforms in the future.

dg0yt commented:
IMO it doesn't make sense to proliferate complex supports expressions from B to multiple A[with-B].
- The tool already  correctly infers the supported platforms of A[with-B] *in normal use*.
- The supported platforms of B are tightly coupled only to B itself.  
  The A often doesn't need to change at all when B supports another platform.

What is new is that feature-testing *in CI* wants to be informed about expected cascades. This might be a little bit too eager ATM. (OTOH we had A falling out of CI testing because someone blocked B from build. This wasn't perfect either.)

BillyONeal commented:
> * The tool already  correctly infers the supported platforms of A[with-B] _in normal use_.

Right. I think for 'unexpected cascade' that should only ever be a ci.feature.baseline.txt edit; there's no reason to block it should the dependency get fixed (or an overlay port added which fixes it).

BillyONeal commented:
@c8ef Did you intend to do that? I was going to try to fix this for you but I see the branch is gone :(

c8ef commented:
> @c8ef Did you intend to do that? I was going to try to fix this for you but I see the branch is gone :(

That's very kind of you, thank you so much!

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into HEAD

BillyONeal committed:
Put things that look related to our build lab in baseline files rather than supports expressions.

wangle: supports expression removed
folly: supports expression changed to 'native' given that the actual build failures seem to be related to that
mvfst: Simplified the supports clause a bit.
Others: Reverted supports expression changes that only appear to be done to be consistent with folly.

BillyONeal commented:
OK, @c8ef , sorry for the running around; we are going to make the feature testing for cascades
less draconian because you're far from the only person who seems ... unhappy with it.

Here are some notes, please let me know if they make sense:

* cachelib: All changes here look good.
* wangle: It looks like this supports expression was added only due to folly getting one, so I
    removed it.
* It looks like most of the other ports all depend on folly, so looking at that next.
* folly's android builds look unrelated to our build lab, so the existing edit to move into
    the supports expression appears to be correct. However, the cause appears cross compiling being
    broken, so I've changed the supports expression to 'native' instead of 'linux | osx':
    ```console
    -- Performing Test FOLLY_HAVE_UNALIGNED_ACCESS
    CMake Error: try_run() invoked in cross-compiling mode, please set the following cache variables appropriately:
       FOLLY_HAVE_UNALIGNED_ACCESS_EXITCODE (advanced)
    For details see /vcpkg/buildtrees/folly/x64-android-dbg/TryRunResults.cmake
    ```
* The other ones that just copied folly's supports clause I removed.
* mvfst: After demorganing the supports clause it becomes !windows | static; I suspect the actual
    value should just be 'static' and we aren't seeing in our lab because we don't test dynamic
    triplets for non-Windows.  
    Ah, nope! `./vcpkg install mvfst --host-triplet x64-linux-dynamic` succeeds! So I just left the
    demorganing.

BillyONeal committed:
Update feature baseline.

LilyWangLL commented:
```
Feature Test [1/5] fizz[core]:x64-android
Skipping testing of fizz:x64-android@2025.04.21.00 because the following dependencies are not supported on x64-android:
folly[core]:x64-android only supports (windows & x64 & !uwp & !mingw) | (native & (x64 | arm64))
Feature Test [2/5] wangle[core]:x64-android
Skipping testing of wangle:x64-android@2025.04.21.00 because the following dependencies are not supported on x64-android:
folly[core]:x64-android only supports (windows & x64 & !uwp & !mingw) | (native & (x64 | arm64))
Feature Test [3/5] mvfst[core]:x64-android
Skipping testing of mvfst:x64-android@2025.04.21.00 because the following dependencies are not supported on x64-android:
folly[core]:x64-android only supports (windows & x64 & !uwp & !mingw) | (native & (x64 | arm64))
Feature Test [4/5] fbthrift[core]:x64-android
Skipping testing of fbthrift:x64-android@2025.04.21.00 because the following dependencies are not supported on x64-android:
folly[core]:x64-android only supports (windows & x64 & !uwp & !mingw) | (native & (x64 | arm64))
Feature Test [5/5] proxygen[core]:x64-android
Skipping testing of proxygen:x64-android@2025.04.21.00 because the following dependencies are not supported on x64-android:
folly[core]:x64-android only supports (windows & x64 & !uwp & !mingw) | (native & (x64 | arm64))
fbthrift[core]:x64-android resulted in the unexpected state cascade because the following dependencies did not build: folly[core]:x64-android only supports (windows & x64 & !uwp & !mingw) | (native & (x64 | arm64))
```

`fbthrift(android) = cascade` need to be added to `scripts/ci.feature.baseline.txt`.

BillyONeal committed:
I missed checking fbthrift, thanks @LilyWangLL!

c8ef commented:
Thank you very much for your prompt feedback! I'm happy with the current situation.

dg0yt commented on ports/folly/vcpkg.json:
```diff
@@ -1,10 +1,10 @@
   "homepage": "https://github.com/facebook/folly",
   "license": "Apache-2.0",
-  "supports": "(windows & x64 & !uwp & !mingw) | (!windows & (x64 | arm64))",
+  "supports": "(windows & x64 & !uwp & !mingw) | (native & (x64 | arm64))",
```
`native` also excludes x64-linux-dynamic or x64-linux-release on x64-linux ("almost-native" in my personal dictionary).

BillyONeal replied:
> `native` also excludes x64-linux-dynamic or x64-linux-release on x64-linux ("almost-native" in my personal dictionary).

Use `--host-triplet x64-linux-dynamic`

dg0yt replied:
I know the worarounds. But this is not a way to sell vcpkg.

BillyONeal replied:
> I know the workarounds. But this is not a way to sell vcpkg.

Do those things not set `CMAKE_CROSSCOMPILING` which would thus trigger this problem?

BillyONeal committed:
Revert the 'native' supports as requested by @dg0yt .

BillyONeal committed:
Also block Android.

BillyONeal commented:
VCPKGMM-06 is unhappy, restarting

BillyONeal commented:
/azp run

BillyONeal commented:
Thanks for the update and sorry again for the spam!
